provider/pkg/provider/environment.go

```diff
@@ -0,0 +1,271 @@
+	}, nil
+}
+
+func (st *PulumiServiceEnvironmentResource) Delete(req *pulumirpc.DeleteRequest) (*pbempty.Empty, error) {
+	input, err := ToPulumiServiceEnvironmentInput(req.GetProperties())
+	if err != nil {
+		return nil, err
+	}
+
+	err = st.client.DeleteEnvironment(context.Background(), input.OrgName, input.EnvName)
+	if err != nil {
+		return nil, err
+	}
+	return &pbempty.Empty{}, nil
+}
+
+func (st *PulumiServiceEnvironmentResource) Create(req *pulumirpc.CreateRequest) (*pulumirpc.CreateResponse, error) {
+	input, err := ToPulumiServiceEnvironmentInput(req.GetProperties())
+	if err != nil {
+		return nil, err
+	}
+
+	// First check if yaml is valid
+	_, diagnostics, err := st.client.CheckYAMLEnvironment(context.Background(), input.OrgName, input.Yaml)
+	if err != nil {
+		return nil, err
+	}
+	if diagnostics != nil {
+		return nil, fmt.Errorf("failed to create environment, yaml code failed following checks: %+v", diagnostics)
+	}
+
+	// Then create environment, and update it with yaml provided. ESC API architecture doesn't let you do it in one call
+	err = st.client.CreateEnvironment(context.Background(), input.OrgName, input.EnvName)
+	if err != nil {
+		return nil, err
+	}
+	diagnostics, err = st.client.UpdateEnvironment(context.Background(), input.OrgName, input.EnvName, input.Yaml, ETAG)
+	if err != nil {
+		return nil, err
+	}
+	if diagnostics != nil {
+		return nil, fmt.Errorf("failed to update brand new environment with pre-checked yaml, due to failing the following checks: %+v \n"+
+			"This should never happen, if you're seeing this message there's likely a bug in ESC APIs", diagnostics)
+	}
+
+	propertyMap, err := input.ToPropertyMap()
+	if err != nil {
+		return nil, err
+	}
+	outputProperties, err := plugin.MarshalProperties(
+		propertyMap,
+		plugin.MarshalOptions{},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pulumirpc.CreateResponse{
+		Id:         path.Join(input.OrgName, input.EnvName),
+		Properties: outputProperties,
+	}, nil
+}
+
+func (st *PulumiServiceEnvironmentResource) Check(req *pulumirpc.CheckRequest) (*pulumirpc.CheckResponse, error) {
+	inputMap, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var failures []*pulumirpc.CheckFailure
+	for _, p := range []resource.PropertyKey{"organization", "name", "yaml"} {
+		if !inputMap[(p)].HasValue() {
+			failures = append(failures, &pulumirpc.CheckFailure{
+				Reason:   fmt.Sprintf("missing required property '%s'", p),
+				Property: string(p),
+			})
+		}
+	}
+
+	return &pulumirpc.CheckResponse{Inputs: req.GetNews(), Failures: failures}, nil
+}
+
+func (st *PulumiServiceEnvironmentResource) Update(req *pulumirpc.UpdateRequest) (*pulumirpc.UpdateResponse, error) {
+	input, err := ToPulumiServiceEnvironmentInput(req.GetNews())
+	if err != nil {
+		return nil, err
+	}
+
+	diagnostics, err := st.client.UpdateEnvironment(context.Background(), input.OrgName, input.EnvName, input.Yaml, ETAG)
```

Comment:
Might want to store the actual ETag in the state and pass it in here. That would be something of a double-edged sword, as it would ensure that manually-edited environments need to be refreshed prior to an update.

Otherwise this should just pass the empty string instead of `ETAG`.

Reply:
Same as above, replaced with empty string